### PR TITLE
feat!: bump all upgradable (dev)Dependencies except pure ESM

### PR DESCRIPTION
**Pull request by JounQin**
close
  - #368
  - #369
  - #370
  - #386
  - #410
  - #416
  - #417
  - #418
  - #422
  - #423
  - #424

**What**:

Maintenance

**Why**:

The (dev)Dependencies are very outdated

**How**:

1. Bump all upgradable (dev)Dependencies except pure ESM ones
2. Upgrade GitHub Actions versions and related usage
3. Enable CI workflow for `pull_request` events
4. Replace [`husky`](https://github.com/typicode/husky) with [`simple-git-hooks`](https://github.com/toplenboren/simple-git-hooks) which is much easier to use, you may need to run `git config --unset core.hooksPath` if you've enabled `husky@v2+` previously
5. Remove `cli-test` task (which is buggy) in favor of `test` including `cli.spec` already, now `npm start validate` is just working as expected
6. Use [eslint-plugin-node-dependencies](https://github.com/ota-meshi/eslint-plugin-node-dependencies) to ensure we have correct versions range of dependencies
7. Enable [`CodeSandbox CI`](https://ci.codesandbox.io) for testing before releasing easily

---

I will start to migrate to pure ESM later

### CODE REVIEW

File: patches/@rushstack+eslint-patch+1.1.4.patch [hunk 1]
```diff
@@ -0,0 +1,13 @@
+diff --git a/node_modules/@rushstack/eslint-patch/lib/modern-module-resolution.js b/node_modules/@rushstack/eslint-patch/lib/modern-module-resolution.js
```
Comment (idahogurl) [inline patches/@rushstack+eslint-patch+1.1.4.patch:1]: I think this is out of scope for this PR. We should create a new PR in the `prettier-eslint` package that addresses the @rushstack/eslint-patch issue reported in #434

File: test/tests/cli.spec.js [hunk 2]
```diff
@@ -11,9 +10,6 @@ const pWriteFile = pify(fs.writeFile);
 const pReadFile = pify(fs.readFile);
 const pUnlink = pify(fs.unlink);
 
-// this is a bit of a long running test...
-jasmine.DEFAULT_TIMEOUT_INTERVAL = 20000;
-
```
Comment (idahogurl) [inline test/tests/cli.spec.js:16]: I'm thinking we should keep the timeout. It may have put there because it was failing CI. Also, not every developer has a really fast computer. It should be replaced with

```js
jest.setTimeOut(20000);
```

File: package.json [hunk 3]
```diff
@@ -12,67 +12,68 @@
...
+    "boolify": "^1.0.1",
+    "camelcase-keys": "^7.0.2",
+    "chalk": "^4.1.2",
+    "common-tags": "^1.8.2",
+    "core-js": "^3.24.1",
+    "eslint": "^8.21.0",
+    "find-up": "^5.0.0",
+    "get-stdin": "^8.0.0",
+    "glob": "^7.2.3",
+    "ignore": "^5.2.0",
```
Comment (idahogurl) [inline package.json:38]: I thought we were only updating devDependencies in this PR.

**Review by idahogurl — CHANGES_REQUESTED** [review]

File: test/tests/cli.spec.js (hunk 2 above)
Comment (JounQin) [reply test/tests/cli.spec.js:16]: Sure.

File: package.json (hunk 3 above)
Comment (JounQin) [reply package.json:38]: I'd like to bump all for a new major release.

File: patches/@rushstack+eslint-patch+1.1.4.patch (hunk 1 above)
Comment (JounQin) [reply patches/@rushstack+eslint-patch+1.1.4.patch:1]: Without this patch, the test will fail, `@rushstack/eslint-patch` is out controlled, and it is only a nested dev dependency via `eslint-config-kentcdodds`, so I don't think we should exclude this part.

Comment (idahogurl) [reply patches/@rushstack+eslint-patch+1.1.4.patch:1]: @JounQin Gotcha. I didn't know it was a dependency of `eslint-config-kentcdodds`

File: package.json [hunk 4]
```diff
@@ -12,67 +12,68 @@
...
-    "lint-staged": "^8.2.1",
-    "nps": "^5.9.5",
+    "@babel/cli": "^7.18.10",
+    "@babel/core": "^7.18.10",
+    "@babel/node": "^7.18.10",
+    "@babel/preset-env": "^7.18.10",
+    "all-contributors-cli": "^6.20.0",
+    "eslint-config-kentcdodds": "^20.3.1",
+    "jest": "^28.1.3",
+    "lint-staged": "^13.0.3",
```
[the remark this answers is not in the document]
Comment (JounQin) [reply package.json]: I'll add [`eslint-plugin-node-dependencies`](https://github.com/ota-meshi/eslint-plugin-node-dependencies) to ensure we have correct dependencies.

File: .nvmrc [hunk 5]
```diff
@@ -0,0 +1 @@
+>=12.22
```
Comment (idahogurl) [inline .nvmrc:1]: This is the version needed by `lint-staged`. I only know that `lint-staged@13` doesn't work on 12 because I'm upgrading the same package on the repos at work. 😆 
```suggestion
>=14.13.1
```

**Review by idahogurl — CHANGES_REQUESTED** [review]

File: .nvmrc (hunk 5 above)
Comment (JounQin) [reply .nvmrc:1]: I don't understand what's the purpose to change it here? We're still supporting `v12` as mentioned above, so the contributor can also use `v12`?

Comment (idahogurl) [reply .nvmrc:1]: The contributor can't use version 12 because `lint-staged@13` won't work. `lint-staged@12.5` will though.

Comment (idahogurl) [reply .nvmrc:1]: https://github.com/okonet/lint-staged/blob/master/package.json

Comment (JounQin) [reply .nvmrc:1]: I've downgraded it in this commit. See [`8902edc` (#437)](https://github.com/prettier/prettier-eslint-cli/pull/437/commits/8902edc917a3e6d71cd0d0dda383ee99bce1c84a#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R55)

**Review by kylemh — APPROVED** [review]
seems fine to me

**Comment by JounQin** [comment]
@kylemh @lydell 

Don you have permissions to install GitHub Apps on this repository? I'd like to install https://ci.codesandbox.io to address https://github.com/prettier/prettier-eslint-cli/pull/424#issuecomment-1200525921

**Comment by kylemh** [comment]
Go for it. We're all in charge here.

**Comment by lydell** [comment]
I don’t have access. I have very few permissions in the repo, it looks like.

**Comment by JounQin** [comment]
cc @prettier/core

**Comment by fisker** [comment]
image
image


Is this what you need?
